### PR TITLE
Developement smart rest upload

### PR DESCRIPTION
**Pull request by mbay-SAG**
Added functionality of uploading smartrest template

### CODE REVIEW

File: c8ydm/agentmodules/smartRest_inizializer.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+class SmartRestInitializer(Initializer):
+    logger = logging.getLogger(__name__)
+
+    def getMessages(self):
+        template_id = 'c8y-dm-agent-v2.0'
+        self.logger.info(f'SmartRest Template Initializer called...')
+        if not self.agent.rest_client.check_SmartRest_template_exists(template_id):
+            self.logger.info(f'SmartRest Template does not exist, creating....')
+            with open('/root/.cumulocity/DM_Agent.json') as f:
```
Comment (switschel) [inline c8ydm/agentmodules/smartRest_inizializer.py:32]: This should be changed to ~ user path as /root/ will not working on windows systems and as non-root users.

File: c8ydm/agentmodules/smartRest_inizializer.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+                payload = f.read()
+                self.logger.debug(f'SmartRest Template readed from file: {payload}')
+            self.agent.rest_client.create_SmartRest_template(payload,template_id)
+            msg = SmartRESTMessage('s/us', '400', ['c8y_SmartRestTemplateUpload', 'C8Y DM Agent Uploaded a SmartRest Template'])
```
Comment (switschel) [inline c8ydm/agentmodules/smartRest_inizializer.py:36]: msg is not used in return []. Should be added there?

File: c8ydm/client/rest_client.py [hunk 3]
```diff
@@ -237,3 +237,59 @@ def set_operations_to_failed(self, operations):
...
+                    self.logger.warning('Got response with status_code: ' +
+                            str(response.status_code))
+                    return False
+            else:
+                self.logger.warning('Response from request: ' + str(response.text))
+                self.logger.warning('Got response with status_code: ' +
+                            str(response.status_code))
+                return False
+        except Exception as e:
+            self.logger.error('The following error occured: %s' % (str(e)))
```
Comment (switschel) [inline c8ydm/client/rest_client.py:273]: More meaningful Error messages please 🗡️

File: c8ydm/client/rest_client.py [hunk 4]
```diff
@@ -237,3 +237,59 @@ def set_operations_to_failed(self, operations):
...
+                self.logger.info('Managed object exists in C8Y')
+                self.logger.debug('Returning the internalID')
+                json_data = json.loads(response.text)
+                return True
+            else:
+                self.logger.warning('Response from request: ' + str(response.text))
+                self.logger.warning('Got response with status_code: ' + str(response.status_code))
+                return False
+        except Exception as e:
+            self.logger.error('The following error occured: %s' % (str(e)))
```
Comment (switschel) [inline c8ydm/client/rest_client.py:294]: More meaningful Error messages.. Example: "Error in Checking Smart Rest Templates..."

**Review by switschel — CHANGES_REQUESTED** [review]
Great work! Found only a few things that should be changed!

File: c8ydm/agentmodules/smartRest_inizializer.py (hunk 1 above)
Comment (mbay-SAG) [reply c8ydm/agentmodules/smartRest_inizializer.py]: did change to ~/.cumulocity

File: c8ydm/client/rest_client.py (hunk 3 above)
Comment (mbay-SAG) [reply c8ydm/client/rest_client.py]: Added more meaningful message

File: c8ydm/client/rest_client.py (hunk 4 above)
Comment (mbay-SAG) [reply c8ydm/client/rest_client.py]: Added more meaningful message

File: c8ydm/agentmodules/smartRest_inizializer.py (hunk 2 above)
Comment (mbay-SAG) [reply c8ydm/agentmodules/smartRest_inizializer.py:36]: I checked back with other implementations using the the msg and copied return []. If not needed I could remove it.

Comment (switschel) [reply c8ydm/agentmodules/smartRest_inizializer.py:36]: The point is why you define a SmartRESTMessage if you don't send it to the platform :-)

Comment (mbay-SAG) [reply c8ydm/agentmodules/smartRest_inizializer.py:41]: done

**Review by switschel — APPROVED** [review]

**Pull request merged**